CMakeLists.txt

```
@@ -328,6 +328,17 @@ endif()
 
 message("FINAL CMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}")
 
+if(${VELOX_ENABLE_GPU})
+  enable_language(CUDA)
+  set(CMAKE_CUDA_FLAGS
+      "-arch=sm_60"
+      "-gencode=arch=compute_60,code=sm_60"
+      "-gencode=arch=compute_60,code=compute_60"
+      "-gencode=arch=compute_70,code=sm_70"
+      "-gencode=arch=compute_70,code=compute_70")
+  list(JOIN CMAKE_CUDA_FLAGS " " CMAKE_CUDA_FLAGS)
```

What is the purpose of `list(JOIN` here? Do you want to add to (potentially) existing `CMAKE_CUDA_FLAGS`? This is just repeating the args you set twice as `set(...` overwrites the existing flags.

To add to existing flags:
```suggestion
  list(APPEND CMAKE_CUDA_FLAGS
      "-arch=sm_60"
      "-gencode=arch=compute_60,code=sm_60"
      "-gencode=arch=compute_60,code=compute_60"
      "-gencode=arch=compute_70,code=sm_70"
      "-gencode=arch=compute_70,code=compute_70")
```

---

This is converting `CMAKE_CUDA_FLAGS` to be separated by space " " instead of by semicolon ";"

---

No not exactly, this is converting the list to a single string with a space as the glue char ([docs](https://cmake.org/cmake/help/latest/command/list.html#join)). Which prevents loops etc. from working properly. 

Why is it necessary to do this? afaik `CMAKE_CUDA_FLAGS` behaves similarly to all the other flag vars and we don't need to convert these?

```
set(a_list
	"a"
	"b")

list(LENGTH a_list al)
message(STATUS "length: ${al}")
list(JOIN a_list " " a_list)
list(LENGTH a_list al)
message(STATUS "length: ${al}")

foreach(e in ${a_list})
message(STATUS "element: ${e}")
endforeach()
```
produces:
```
-- length: 2
-- length: 1
-- element: a b
``

---

Because the compiler will complain them being joined with semicolon.  Is there a way to specify a list variable being joined using space when passed to compiler?

---

Hm, interesting I haven't used cmake with cuda yet so I'll have to look into it. I'll come back to this soon

---

@Yuhta could you try this diff? It uses a feature introduced in 3.18 that automatically determines the correct flags. For me it generates these ` --generate-code=arch=compute_52,code=[compute_52,sm_52] ` (which is obv to low to compile this test)

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index d1d242b97..4ab8af4c7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -329,14 +329,9 @@ endif()
 message("FINAL CMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}")
 
 if(${VELOX_ENABLE_GPU})
+  # determine CUDA_ARCHITECTURES automatically
+  cmake_policy(SET CMP0104 NEW)
   enable_language(CUDA)
-  set(CMAKE_CUDA_FLAGS
-      "-arch=sm_60"
-      "-gencode=arch=compute_60,code=sm_60"
-      "-gencode=arch=compute_60,code=compute_60"
-      "-gencode=arch=compute_70,code=sm_70"
-      "-gencode=arch=compute_70,code=compute_70")
-  list(JOIN CMAKE_CUDA_FLAGS " " CMAKE_CUDA_FLAGS)
 endif()
 
 set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
```

---

(also if we want to set compile options we should be able to use `target_compile_options` using the modern style as CUDA has full language support)

---

Emm I am getting `CUDA_ARCHITECTURES is empty for target "velox_gpu_hash_table_test"` with cmake 3.27.1

---

Even with the policy set? Weird...